[RFC]: Integrate assertions functionalization to export (after AOT export)

XuanQi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104203
* __->__ #103887
* #103757
* #103751

This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.

XuanQi committed:
[RFC]: Integrate assertions functionalization to export (after AOT export)

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/functionalize_assertions.py:
```diff
@@ -12,7 +14,15 @@
 
-def functionalize(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
+@dataclass
+class FunctionalizedGraphModule:
```
It should rename it to something else as naming it GraphModule can create a confusion about whether it has same functionality as normal GraphModule.

XuanQi replied:
I guess nothing is harder than naming so let me just return a tuple for simplicity :p

tugsbayasgalan commented on torch/_export/exported_program.py:
```diff
@@ -164,10 +167,115 @@ def transform(self, *passes: PassType) -> "ExportedProgram":
+    )
+
+
+def _update_after_adding_runtime_assertions(
```
Why do we need this utility function? I feel like after transform, we should always produce a valid ExportedProgram with correct ExportGraphSignature etc.

XuanQi replied:
This is mostly a fix for current logic (I might misunderstand [`FQN`](https://github.com/pytorch/pytorch/blob/8ce4fee68d77c6c820837d105ae36ed34bbc8167/torch/_functorch/aot_autograd.py#L939) in [`GraphSignature`](https://github.com/pytorch/pytorch/blob/8ce4fee68d77c6c820837d105ae36ed34bbc8167/torch/_functorch/aot_autograd.py#LL942C7-L942C21)) - what I've seen is that after adding runtime assertions, the FQNs for output related fields will change so the previous graph signature copied after AOT export (https://github.com/pytorch/pytorch/blob/8ce4fee68d77c6c820837d105ae36ed34bbc8167/torch/_export/__init__.py#L200) is incorrect. Use the example I have in test plan:
```
buf = torch.ones(6, 2)
weight = 0.01
bias = 0.2
d_in = 3
d_out = 4
class Foo(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.register_buffer("buf", buf)

        self.linear = torch.nn.Linear(d_in, d_out)
        self.linear.weight.data.fill_(weight)
        self.linear.bias.data.fill_(bias)

    def forward(self, x):
        self.buf.add_(5)
        return self.linear(x).cos() + self.buf.sum()
```
Without this change, the graph signature will be something like:
```
ExportGraphSignature(
    ...
    user_outputs=['add_1'],
    buffers_to_mutate={'add': 'L__self___buf'},
    ...
)
```
And the graph is like:
```
graph():
    %arg0_1 : [#users=3] = placeholder[target=arg0_1]
    %arg1_1 : [#users=2] = placeholder[target=arg1_1]
    %arg2_1 : [#users=3] = placeholder[target=arg2_1]
    %arg3_1 : [#users=3] = placeholder[target=arg3_1]
    ...
    %add_tensor_1 : [#users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%cos_default, 
    %sum_default), kwargs = {})
    return (add_tensor, add_tensor_1)
```
So as you could see, `user_outputs` and `buffers_to_mutate` is incorrect now since the output names are `add_tensor` and `add_tensor_1` now instead of `add` and `add_1` (note this is already before the change of this PR).

tugsbayasgalan commented on test/export/test_funtionalized_assertions.py:
```diff
@@ -112,6 +120,88 @@ def f(x):
+                inputs_to_buffers={"arg2_1": "L__self___buf"},
+                buffers_to_mutate={"add_tensor": "L__self___buf"},
+                backward_signature=None,
+                assertion_dep_token_output="dep_token_8",
```
Maybe also better to add the position of the assertion_dep_token_output as well?

XuanQi replied:
I changed to use the dep_token output name instead of index since:
* It seems more robust to use name compared to index.
* It's a little bit redundant to have both so I removed index.

But I can add it back if preferred as planned previously.

tugsbayasgalan commented on test/export/test_funtionalized_assertions.py:
```diff
@@ -112,6 +120,88 @@ def f(x):
+                (buf.add(5), Foo()(inp), _DEP_TOKEN),
+            )
+        )
+        self.assertEqual(ep(inp)[1], _DEP_TOKEN)
```
This should work:
`self.assertEqual(ep(inp), eager_function(inp))`

tugsbayasgalan commented on torch/_export/exported_program.py:
```diff
@@ -164,10 +167,115 @@ def transform(self, *passes: PassType) -> "ExportedProgram":
+    # within a tuple if the exported program is called directly. For example,
+    # if original output is `(x, y)`, the output with dep token included will be
+    # like `((x, y), dep_token)`.
+    out_spec = pytree.TreeSpec(
```
We need to preserve in_spec and out_spec.

tugsbayasgalan commented on test/export/test_funtionalized_assertions.py:
```diff
@@ -112,6 +120,88 @@ def f(x):
+
+            def forward(self, x):
+                self.buf.add_(5)
+                return self.linear(x).cos() + self.buf.sum()
```
Can you add a test where you add the python assertion like `assert x.shape[0] > 4` or sth in the code and see it also picks up the dep token?

XuanQi replied:
Added the test case `test_native_python_assertion`. However, I still get issue (maybe it's expected though) as mentioned last week:
* If I don't specify input constraints of export, the assertion won't appear in the exported graph (instead I got `_functional_assert_async.msg` like `arg0_1.shape[0] is specialized at 2...`, which is expected)
* If I manually specify dynamic constrains like `constraints=[dynamic_dim(inp, 0), dynamic_dim(inp, 1)]`, it will actually fail me with error message `1. Could not validate (strict) constraint 2 <= L['x'].size()[0] <= oo as we generated a guard on this size variable: L['x'].size()[0] > 3.` as shown in screenshot.
And suggest me to specify input constraints as `[3 < dynamic_dim(x, 0),  dynamic_dim(x, 1)]`. This became the same as `assert x.shape[0] > 3` and it passed after I change input constraints as suggested.
![Screenshot 2023-06-20 at 10 50 37 PM](https://github.com/pytorch/pytorch/assets/4644550/53f84914-308f-4498-8650-311013b7c385)

But anyway,  I still add the test case here but in general I would say it's not really testing a new case since:
* It doesn't really covert `assert x.shape[0] > 3` to something like a inline constraint like `constrain_as_value(x.shape[0], min=4)`.

* Instead it tries to:
  * Use specified shape assertions based on example inputs during export.
  * Enforce user to set input constraints which is equivalent to native python assertion.

XuanQi replied:
It turns out this is caused by pytest assertion rewrite. Fixed by adding `PYTEST_DONT_REWRITE ...` comment.

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

tugsbayasgalan commented on test/export/test_funtionalized_assertions.py:
```diff
@@ -101,6 +114,115 @@ def f(x):
+                inputs_to_buffers={"arg2_1": "L__self___buf"},
+                buffers_to_mutate={"add_tensor": "L__self___buf"},
+                backward_signature=None,
+                assertion_dep_token_output="dep_token_8",
```
Maybe we can make this a dictionary like {2: "dep_token_8"} so it is little more concise?

tugsbayasgalan commented on test/export/test_funtionalized_assertions.py:
```diff
@@ -101,6 +114,115 @@ def f(x):
+        )
+        output_node = next(n for n in ep.graph.nodes if n.op == "output")
+        self.assertEqual(
+            [str(arg) for arg in output_node.args[0]],
```
Would be good to assert that len(output_node) == len(user_outputs) + 1 (for assertion token) + len(buffers_to_mutate)

tugsbayasgalan commented on torch/_export/exported_program.py:
```diff
@@ -164,10 +191,106 @@ def transform(self, *passes: PassType) -> "ExportedProgram":
+    # TODO: Improve current pass infra to make it possible to update graph
+    # signature as well (currently only graph module).
+
+    def _get_output_FQNs(gm: torch.fx.GraphModule) -> List[FQN]:
```
maybe just name them output_node_names?

tugsbayasgalan commented on torch/_export/exported_program.py:
```diff
@@ -164,10 +191,106 @@ def transform(self, *passes: PassType) -> "ExportedProgram":
+            ep = _update_after_adding_runtime_assertions(self, ep)
+            ep = _update_after_functionalizing_runtime_assertions(
+                ep,
+                _functionalize_side_effectful_ops(gm=ep.graph_module),
```
If you make `_functionalize_side_effectful_ops` similar to `_AddRuntimeAssertionsForConstraintsPass`, then you can just have utility function that updates the ExportedProgram signatures.

XuanQi replied:
So I think:
* Ideally for each pass, the signature update should follow graph transformation.
* However, I don't see easy way to implement the above using fx `PassManager`.
* A better solution I think is to just create a `ExportPassManager` which allowing running signature update after each pass run.

But I think this could be a separate PR and goes beyond the scope? But anyway, let me at least add `update_exported_program_signature ` as a class instance method of base class `ExportPassBase` so each concrete could override it (even we need to call it manually for now).

tugsbayasgalan commented on torch/_export/exported_program.py:
```diff
@@ -164,10 +191,106 @@ def transform(self, *passes: PassType) -> "ExportedProgram":
+            # Ideally `_update_after_adding_runtime_assertions` should run
+            # whenever `_AddRuntimeAssertionsForConstraintsPass` runs, here
+            # bundle it with `functionalize_assertions` for safety.
+            ep = _update_after_adding_runtime_assertions(self, ep)
```
This should also run outside of this if condition right. Since it is a bug that the signature is not updated, we should also run this before the if statement.

XuanQi replied:
Yes. As mentioned in the comment I added there just for safety but if we're OK with moving it out, let me just move it out then.

tugsbayasgalan reviewed (commented):
Mostly looks good :) Thanks for pushing through this. My only comments now are just code refactoring, I think we want to:
1. Make `functionalize_side_effectful_ops` an interpreter based pass similar to `_AddRuntimeAssertionsForConstraintsPass`
2. Use same utility function for both `_AddRuntimeAssertionsForConstraintsPass` and `functionalize_side_effectful_ops` to update the exported graph signature. 

Let me know what you think.

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

avikchaudhuri commented on test/export/test_passes.py:
```diff
@@ -414,6 +420,118 @@ def f(x):
             "torch.ops.aten._assert_async.msg", 0, exactly=True
         ).run(gm.code)
 
+    def test_functionalization(self) -> None:
```
love the tests!

avikchaudhuri commented on torch/_export/__init__.py:
```diff
@@ -235,7 +236,9 @@ def export(
             if _add_runtime_assertions:
-                exported_program = exported_program._add_runtime_assertions()
+                exported_program = exported_program._add_runtime_assertions(
+                    functionalize_assertions=_functionalize_runtime_assertions,
```
can shorten `functionalize_assertions` arg to simply `functionalize` (because we already call `_add_runtime_assertions`)

avikchaudhuri commented on torch/_export/__init__.py:
```diff
@@ -120,6 +120,7 @@ def export(
     constraints: Optional[List[Constraint]] = None,
     *,
     _add_runtime_assertions=True,
+    _functionalize_runtime_assertions=False,
```
The reason this flag exists is that we have a different signature with it, right? What would be the plan to turn it on?

Without this flag, how broken is the current state again? Will the assertion op still be DCE-d before future transforms can run? @tugsbayasgalan @angelayi can you remind me of the overall rollout plan here?

XuanQi replied:
So this flag is mostly for gating purpose before we could turn on this feature by default. But yeah I guess it's worth to discuss rollout plan as well. cc @tugsbayasgalan

> Without this flag, how broken is the current state again
My rough understanding is nothing is broken due to DCE since https://github.com/pytorch/pytorch/blob/618cc82e770cf758bc9cbcdfd4c61e2d73bdca25/torch/fx/node.py#L33-L41?

avikchaudhuri reviewed (approved):
code lgtm, left some questions

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:
```diff
@@ -51,6 +56,41 @@ def _convert_range_to_int(range: RangeConstraint):
 
 class _AddRuntimeAssertionsForConstraintsPass(ExportPassBase):
+    @classmethod
+    def update_exported_program_signature(
```
Let's move this back to exported_program.py. I feel little uneasy about exposing this as a class method and introducing a circular dependency. As a follow up, we will discuss what the proper way to abstract the passes that can work with ExportedProgram.

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Integrate assertions functionalization to export (after AOT export)"


This PR integrated the assertion functionalization logic into current export logic.

**NOTE:**
I finally decided to do the assertion functionalization after AOT export instead of before for the following reasons:
* The benefit of AOT export is that the graph is already functionalized so things like method call is already transformed to function call. However, if we do it before AOT export, the graph is still in torch level and extra logic like https://github.com/pytorch/pytorch/blob/bab21d20ebf45a5dc620b48791bb526f664445a5/torch/_export/pass_base.py#L201-L204C17 will need to be implemented.
* The graph signature is kind of already incorrect after adding runtime assertions currently (this doesn't seem break logic since we already depend on positions instead of FQNs of outputs). This PR also fixed this.



[ghstack-poisoned]

XuanQi commented:
@pytorchbot merge

XuanQi commented:
@pytorchbot label "topic: not user facing"

tugsbayasgalan commented on torch/_export/exported_program.py:
```diff
@@ -166,10 +191,106 @@ def transform(self, *passes: PassType) -> "ExportedProgram":
+        # separate step. However this requires augmenting pass infra at fx level
+        # to operate on `ExportedProgram` instead of `fx.GraphModule`.
+        # TODO: Integrate graph signature update into pass run.
+        ep = _update_graph_signature_after_adding_runtime_assertions(
```
If I remember correctly, this can happen after any pass right? Should we rename this to something more generic? (like fixup_graph_signature)

XuanQi replied:
Will handle in a separate small patching PR :)
